watertap/unit_models/zero_order/nanofiltration_zo.py

```diff
@@ -0,0 +1,71 @@
+###############################################################################
+# WaterTAP Copyright (c) 2021, The Regents of the University of California,
+# through Lawrence Berkeley National Laboratory, Oak Ridge National
+# Laboratory, National Renewable Energy Laboratory, and National Energy
+# Technology Laboratory (subject to receipt of any required approvals from
+# the U.S. Dept. of Energy). All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. These files are also available online at the URL
+# "https://github.com/watertap-org/watertap/"
+#
+###############################################################################
+"""
+This module constians a ero-order representation of a nanofiltration unit
+operation.
+"""
+
+from pyomo.environ import Constraint, units as pyunits, Var
+from idaes.core import declare_process_block_class
+
+from watertap.core.zero_order_sito import SITOBaseData
+
+# Some more inforation about this module
+__author__ = "Andrew Lee"
+
+
+@declare_process_block_class("NanofiltrationZO")
+class NanofiltrationZOData(SITOBaseData):
```

Comment:
We do have a zero-order NF model in `unit_models` already but I'm not sure whether you intended to focus on an NF model based on the `SITOBaseData` class. Since we eventually want WT3/Proteus models to be compatible on flowsheets, it might be good to try this with the NF model that we have now (as an early first step).

Reply:
The purpose where is to demonstrate the standard workflow for all models that will come from WT3, so it needs to use `SITOBaseData`. All things considered, I would suggest it would be better to look to see if the existing zero-order NF model could be replaced by this than the other way around.

Reply:
The two main variables that the existing zero-order NF model includes (that the SITOBase NF doesn't) are membrane area and membrane water (solvent) flux.

Reply:
@TimBartholomew might want to jump in and comment on anything else that the existing zero-order NF model includes that should be incorporated into the SITOBase NF model.

Reply:
Is membrane area used to determine performance, or estimated based on the flow of water to be treated? If it is the former, then that is becoming a predictive model (even if very simple) rather than what we are calling zero-order models here. If it is the later, then we can look at adding it but we would need an area/flow parameter or similar.

Reply:
So the volumetric water flux, which is the m3/s/m2 of membrane area, would be fixed, and then either the membrane area is fixed, or the volumetric recovery rate is fixed. Thus, I think this would fall under the latter case that you mentioned.

Another thing to note is the existing NF_ZO model was modified a little bit to be compatible with the generic property framework (tested with `seawater_ion_generic`). Before those modifications, the model did not work with the generic property framework as written. Just another consideration before we think about replacing the existing zero-order NF.

Reply:
The problem with the first point is it now means that the structure of the NF model would be different to all the other, and require a different parameter for recovery - not a show stopper but more things that need to be documented. Also, the current database on has the overall recovery parameter, not the recovery per unit area.

For the second, the new model should work for any property package that defines volumetric flow and mass based concentrations - perhaps not the most generic of properties but that is the documented foundation of the WT3 approach. If the ProteusLib zero-order NF model needed special handling then that implies you added something unusual to that model.

Reply:
The overall (volumetric) recovery is also in the Proteus 0-order NF model--it just has membrane water flux, a typical and often used variable, and area too. We can add those two vars to this NF model without altering recovery. All we'd do is additionally fix the membrane water flux after building. @kurbansitterley - do you think it'd be reasonable to augment the NF model like this?

On the second point, yes- I would say I did something unusual for the Proteus 0-order NF because the way it was written before, it could only handle mass flow basis, not mole, and it only detected solutes, not ions (if I am recalling correctly). If you want to have a quick look, check out PR #238 or you could look directly at the [NF_ZO file](https://github.com/adam-a-a/watertap/blob/c9c3cfee95244e719d94926c311ed00f582eee4a/watertap/unit_models/nanofiltration_ZO.py).

Reply:
With all that said, I think we could initially keep both versions until we make a decision on how to merge them. Ultimately, since we will eventually try to make WT3/Plib models interchangeable in flow sheets, WT3 models will need added functionality anyway, so merging capabilities of the two models makes sense (even though we'd be jumping the gun at this point).

Reply:
@adam-a-a I don't see anything wrong with this augmentation to the NF model, especially considering your comment on needing to add functionality to WT3 models.

Reply:
@adam-a-a this base class is for zero-order models that are similar to each other (hopefully most models in WT3), I don't think we can make a class that covers all zero-order models and the ProteusLib zero-order nanofiltration model is a good example of that. That model has a variety of things that are specific to membrane units, like flux and membrane area. We do not want those membrane specific things to be in a generic base class that we want to apply to all different types of technologies. We could have the ProtuesLib zero-order nanofiltration model inherit from this base class and add the membrane specific parameters, a key point I think you made though is if we are locking ourselves into a specific basis (i.e. mass).

@andrewlee94 are there any issues with your current class to providing flexibility for the basis (e.g. mass vs molar)?

Reply:
@TimBartholomew Whilst I think there are no fundamental issues with switching between mass and mole basis, we would need to be careful to ensure that the two do not get mixed up. Currently, I believe everything is done on a mass basis to keep things simple.

Note that the concentration basis is basically determined by the feedwater definition (a yet to be added yml file and Feed block), whilst the removal basis is defined by the unit-specific yml file. We are already allowing for people to mix-and-match databases if desired, so we would need to have checks for the flow basis if we want to allow for both mass and mole bases.